File: floor.py
def generate_rows(width):
    if width == 0:
        return [[]] 
    elif width == 1:
        return [] 

    rows = []
    if width - 2 >= 0:
        rows.extend([row + [2] for row in generate_rows(width - 2)])

    if width - 3 >= 0:
        rows.extend([row + [3] for row in generate_rows(width - 3)])
    return rows

def generate_prefix_sum(row):
    prefix_sum = set()
    prefix = 0
    for i in range(len(row) - 1):
        prefix += row[i]
        prefix_sum.add(prefix)
    return prefix_sum

def generate_conflict_list(prefix_sums):
    conflict_list = []
    for i in range(len(prefix_sums)):
        conflict_list.append([j for j in range(len(prefix_sums)) if prefix_sums[i].isdisjoint(prefix_sums[j])])
    return conflict_list


def calculate_total_ways(conflict_list, height):
    num_rows = len(conflict_list)
    dp = [1] * num_rows  # Base case: 1 way to build the floor with 1 row

    # Build the solution for each height from 2 to the given height
    for _ in range(1, height):
        new_dp = [0] * num_rows
        for i in range(num_rows):
            for j in conflict_list[i]:
                new_dp[j] += dp[i]
        dp = new_dp

    return sum(dp)  # Sum of all ways for the topmost row

def count_patterns(width, height):
    rows = generate_rows(width)
    prefix_sums = [generate_prefix_sum(row) for row in rows] 
    conflict_list = generate_conflict_list(prefix_sums)
    return calculate_total_ways(conflict_list, height)

print(count_patterns(32, 10))

File: c-sharp/Floor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;


namespace FloorProblem
{
    class Floor
    {
        static void Main(string[] args)
        {
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();
            var rows = GenerateRows(32); 

            var prefixSumsList = new List<HashSet<long>>();
            foreach (var row in rows)
            {
                var prefixSums = GeneratePrefixSum(row);
                prefixSumsList.Add(prefixSums);
            }

            var conflictList = GenerateConflictList(prefixSumsList);

            long totalWays = CalculateTotalWays(conflictList, 10);
            stopwatch.Stop();
            long elapsedMilliseconds = stopwatch.ElapsedMilliseconds;
            Console.WriteLine($"Total ways to build the floor: {totalWays}");
            Console.WriteLine($"Elapsed Time: {elapsedMilliseconds} milliseconds");
        }
        static List<List<long>> GenerateRows(long width)
        {
            if (width == 0)
            {
                return new List<List<long>> { new List<long>() };
            }
            if (width == 1)
            {
                return new List<List<long>>();
            }

            List<List<long>> rows = new List<List<long>>();

            if (width - 2 >= 0)
            {
                foreach (var row in GenerateRows(width - 2))
                {
                    var newRow = new List<long>(row);
                    newRow.Add(2);
                    rows.Add(newRow);
                }
            }

            if (width - 3 >= 0)
            {
                foreach (var row in GenerateRows(width - 3))
                {
                    var newRow = new List<long>(row);
                    newRow.Add(3);
                    rows.Add(newRow);
                }
            }

            return rows;
        }
        static HashSet<long> GeneratePrefixSum(List<long> row)
        {
            HashSet<long> prefixSum = new HashSet<long>();
            long prefix = 0;

            for (int i = 0; i < row.Count - 1; i++)
            {
                prefix += row[i];
                prefixSum.Add(prefix);
            }

            return prefixSum;
        }
        static List<List<long>> GenerateConflictList(List<HashSet<long>> prefixSumsList)
        {
            var conflictList = new List<List<long>>();

            for (int i = 0; i < prefixSumsList.Count; i++)
            {
                var conflicts = new List<long>();
                for (int j = 0; j < prefixSumsList.Count; j++)
                {
                    if (i != j && !prefixSumsList[i].Overlaps(prefixSumsList[j]))
                    {
                        conflicts.Add(j);
                    }
                }
                conflictList.Add(conflicts);
            }

            return conflictList;
        }
        static long CalculateTotalWays(List<List<long>> conflictList, long height)
        {
            int numRows = conflictList.Count;
            long[] possibleWays = new long[numRows];
            Array.Fill(possibleWays, 1L);  // Base case: 1 way to build the floor with 1 row

            for (int h = 1; h < height; h++)  // Loop index can stay as int if height will not exceed int.MaxValue
            {
                long[] newPossibleWays = new long[numRows];
                for (int i = 0; i < numRows; i++)  // Loop index can stay as int since numRows is based on conflictList.Count
                {
                    foreach (long j in conflictList[i])
                    {
                        newPossibleWays[j] += possibleWays[i];
                    }
                }
                possibleWays = newPossibleWays;
            }

            long totalWays = 0;
            foreach (long way in possibleWays)
            {
                totalWays += way;
            }
            return totalWays;
        }

    }
}
